app/app_kas.c

```diff
@@ -596,37 +652,105 @@ ACVP_KAS_IFC_TC *tc = NULL;
                 printf("Error encapsulating in KAS-IFC\n");
                 goto err;
             }
-            memcpy_s(tc->ct_z, KAS_IFC_MAX, encap_s, encap_s_len);
-            memcpy_s(tc->pt_z, KAS_IFC_MAX, z, z_len);
-            tc->ct_z_len = (int)encap_s_len;
-            tc->pt_z_len = (int)z_len;
+            memcpy_s(tc->iut_ct_z, KAS_IFC_MAX, encap_s, encap_s_len);
+            memcpy_s(tc->iut_pt_z, KAS_IFC_MAX, z, z_len);
+            tc->iut_ct_z_len = (int)encap_s_len;
+            tc->iut_pt_z_len = (int)z_len;
+        }
+
+        /* For KAS2 initiator tests, regardless of role, we decapsulate the server Z */
+        if (tc->scheme == ACVP_KAS_IFC_KAS2) {
+            if (z) free(z);
```

Comment:
do you need to set z to null?  looks like you could go to err which tries to free z (again) ?